lib/completion-templates.ts

```diff
@@ -44,6 +44,10 @@ _{{app_name}}_yargs_completions()
   IFS=$si
   _describe 'values' reply
 }
-compdef _{{app_name}}_yargs_completions {{app_name}}
+if [[ $zsh_eval_context == *func ]]; then
```

Comment:
This looks a bit magical. The possible values matching `*func` are `evalautofunc `, `loadautofunc`, and `shfunc`.

Is there a reason to test for anything other than `loadautofunc`? Which would be clearer.

Reply:
It's `loadautofunc` when first autoloaded and `shfunc` on subsequent calls.

Reply:
After it gets autoloaded with context `loadautofunc` , I do not think these lines are executed again.

The completion function (not the whole file) gets loaded into shell and called multiple times with context `shfunc`, but I don't think that is relevant to the code outside the completion function.

Reply:
It'e actually the whole file that becomes the completion function unless the file defines a function with the exact same name as the one it was autoloaded with. Which is not the case here ATM. But we can change the function name if you want to go that route.

Reply:
I did carefully save the completion to a file with the name of the autocomplete function, as per advice. I'll try using a different name to check behaviour. I had a small battle to get autoload working but think I have the hang of it now...

Reply:
> It'e actually the whole file that becomes the completion function unless the file defines a function with the exact same name as the one it was autoloaded with. 

Yes. 

> Which is not the case here ATM. But we can change the function name if you want to go that route.

I might be missing something (again). This is the case ATM as long as user creates the file with the name of the function supplied by Yargs?

But I see that if user uses a different filename, the `shfunc` context comes into play. The autoload defines a function based on the filename, which calls/includes the Yargs supplied function. I am slightly concerned `shfunc` could trigger incorrectly in some context outside autoload, but not very likely.

I was thinking the user instructions could mention the autoload case with the "correct" filename, something like:

```
# Installation: foo completion >> ~/.zshrc
#    or foo completion >> ~/.zprofile on OSX.
#    or save completion into executable file named _foo_yargs_completions in autoload fpath
```

Reply:
It would still be safer to have the if like that, it will work either way in that case. It is customary to call the autoload scripts just `_{cmdname}`, e.g. `_git`, having it require a specific name might just cause mistakes for users or package maintainers in the future...